lib/pbench/agent/results.py

```diff
@@ -123,60 +206,143 @@ def make_result_tb(self) -> str:
-                tarball.unlink()
-            except Exception as exc:
-                if not isinstance(exc, OSError) or exc.errno != errno.ENOENT:
-                    raise RuntimeError(
-                        f"Error removing failed tarball, '{tarball}', {exc}"
+            raise RuntimeError(msg)
+        else:
+            if tar_proc.returncode == 0:
+                if xz_proc is not None and xz_proc.returncode != 0:
+                    msg = self._unlink_tarball(
+                        tarball,
+                        f"Failed to create tar ball; 'xz' return code: {xz_proc.returncode:d}",
                     )
+                    raise RuntimeError(msg)
+                else:
+                    try:
+                        e_file.unlink()
+                    except FileNotFoundError:
+                        pass
+                    except Exception as unlink_exc:
+                        self.logger.warning(
+                            "Failed to remove 'tar' stderr file, %s: '%s'",
+                            e_file,
+                            unlink_exc,
+                        )
+            else:
+                # We explicitly ignore the return code from the optional 'xz' process.
+                msg = self._unlink_tarball(
+                    tarball,
+                    f"Failed to create tar ball; 'tar' return code: {tar_proc.returncode:d}",
+                )
+                raise RuntimeError(msg)
+        try:
+            (tar_len, tar_md5) = md5sum(tarball)
+        except Exception:
+            msg = self._unlink_tarball(
+                tarball,
+                f"Failed to verify and generate MD5 for created tar ball, '{tarball}'",
+            )
+            raise RuntimeError(msg)
 
         # The contract with the caller is to just return the full path to the
-        # created tarball.
-        return str(tarball)
+        # created tar ball.
+        return TarballRecord(tarball, tar_len, tar_md5)
 
 
 class CopyResultTb:
-    """CopyResultTb - Use the server's HTTP PUT method to upload a tarball
+    """Interfaces for copying result tar balls remotely using the server's HTTP
+    PUT method for uploads.
     """
 
+    class FileUploadError(Exception):
+        """Raised when the uploading of file to server has failed"""
+
+        pass
+
     def __init__(
-        self, controller: str, tarball: str, config: PbenchAgentConfig, logger: Logger
+        self,
+        controller: str,
+        tarball: str,
+        tarball_len: int,
+        tarball_md5: str,
+        config: PbenchAgentConfig,
+        logger: Logger,
     ):
-        """CopyResultTb contructor - raises FileNotFoundError if the given
-        tar ball does not exist, and a ValueError if the given controller is
-        not a valid hostname.
+        """Contructor for object representing tar ball to be copied remotely.
+
+        Raises
+            ValueError          if the given controller is not a valid hostname
```

Comment:
I don't see where ValueError can be raised...did I miss it?  (Isn't it fun maintaining these lists?  ;-) )